### ISSUE TITLE
replaceComponent is not entirely replacing

### ISSUE THREAD

**Issue by fabien-h** [reactionimp/icemachinesplus#116]
We have a custom navbar container an component : `/Users/fabien/projects/icemachinesplus/imports/plugins/custom/custom-navbar/client/container/navbar.js`

It ends with that code : `replaceComponent("NavBar", NavBar, composeWithTracker(composer));`

But the container is not fully replaced.

If you add a log in the composer function in : `/Users/fabien/projects/icemachinesplus/imports/plugins/core/ui-navbar/client/containers/navbar.js`

It is still logged.

**Comment by jshimko** [comment]
The solution to this is usually to grab the UI component (without its container), then wrap that component with your own container(s) and register a newly named component and use that instead of `NavBar`.  However, doing that also means you are not using any of the data or event handlers in the original container and the UI may actually need some of that to function.

For example, your custom navbar container...
https://github.com/reactionimp/icemachinesplus/blob/master/imports/plugins/custom/custom-navbar/client/containers/navbar.js

...doesn't inject several of the props that the official navbar container does.
https://github.com/reactionimp/icemachinesplus/blob/master/imports/plugins/core/ui-navbar/client/containers/navbar.js

So if you completely remove the official container, not only do you not get any container updates that may happen in the future for that component, but you also remove some of the core navbar functionality that you probably need.  So I think the solution here is probably to continue to inherit the original container, but solve the flickering issue you were seeing.  

For one, it's a definitely a best practice in those composer functions to check the ready state of any subscriptions you're depending on.  Components will jump back and forth between ready state and show a loader when not ready.  You generally want to force the component to not render until all data it depends on is available.  In your case, your container is doing a `Shops.findOne()` [here](https://github.com/reactionimp/icemachinesplus/blob/master/imports/plugins/custom/custom-navbar/client/containers/navbar.js#L9).  You should have a subscription readiness check right before that so it doesn't go any further until that data is definitely ready.  See [this section](https://github.com/reactionimp/icemachinesplus/blob/master/imports/plugins/core/ui-navbar/client/containers/navbar.js#L9-L12) of the original navbar container for an example of what I mean.  Try adding that block of code to yours and see if that fixes the re-rendering issues.

**Comment by fabien-h** [comment]
Yeah, the issue is precisely the inheritance. It means I can only add behaviour and data and not remove. And in our case, I want to remove.

"Components will jump back and forth between ready state and show a loader when not ready"

This is what we do not want. The data are injected at start so there should not be "not ready" moments.

There is no possibility to solve the flickering with the original container. When the data are loading, it displays a loader ; and I cannot overwrite that.

**Comment by jshimko** [comment]
Yeah, there are a couple ways you could probably work around that.  I think what we probably need to do is extend the `replaceComponent` API to allow easily overwriting the containers.  In the meantime, one easy solution would be to use `registerComponent` instead of `replaceComponent`.  Literally the only difference between the two of them is that `replaceComponent` merges previous containers with any new containers and `registerComponent` completely overwrites whatever exists under the provided component name.

Just for reference, here's the code for `replaceComponent`.  Note that it essentially just merges any new containers, and then runs `registerComponent` with the results.  You could skip that by just running `registerComponent` directly with your component/container.

https://github.com/reactioncommerce/reaction/blob/master/imports/plugins/core/components/lib/components.js#L101-L122

**Comment by fabien-h** [comment]
I tried to register my Navbar container and component and to use them as raw and not as extentions of the core ones. But could not figure out how to overwrite the Navbar component in the layout without touching some core code.

In the mean time I just edited the core container navbar.

=> Not a problem right now ; but we cannot update properly now.

**Comment by fabien-h** [comment]
`registerComponent` works. My formulation was bad, sorry.

My issue is : I have my newly registered component and I cannot manage to make reaction use it as the base navbar.

**Comment by jshimko** [comment]
So it sounds like `registerComponent` _doesn't_ work for you then?  You should be able to replace the core navbar and all of its containers with this...

```js
registerComponent("NavBar", CustomNavBar, CustomNavBarContainer);
```

That should completely overwrite whatever is there from core code.  Is that what you tried?

**Comment by jshimko** [comment]
So I tested this myself and I'm not seeing any issues using `registerComponent` to replace everything.  However, you also can't just replace the entire NavBar container without adding some of that logic back in.  That container is responsible for passing in a bunch of data that the NavBar component relies on.  So you probably need to copy the container into your plugin, make the changes you need, then use `registerComponent` to overwrite the original component/container.

I'm also noticing that you've done some extra duplication with the NavBar component that you can skip.  If all you're doing is adding some custom methods/handlers and overwriting the `render` method, you can `extend` the original `NavBar` class and just add the things that are being changed for your custom version.  You'd use `getRawComponent` to do that like this...

```js
const NavBar = getRawComponent("NavBar");

class CustomNavBar extends NavBar {
  customHandler() {
    // custom logic
  },
   
  render() {
    return (
      <div>
        customized render method...
      </div>
    )
  }
}

replaceComponent("NavBar", CustomNavBar);

```

Then you don't need to copy/paste the entire `NavBar` component from core and modify that.  That also allows you to get future updates to the core `NavBar` without having to manually copy them over to your custom copy.
